app/src/main/java/ch/epfl/sdp/contamination/ConcreteCachingDataSender.java

```diff
@@ -46,33 +56,42 @@ void setInteractor(GridFirestoreInteractor interactor) {
 
     @Override
     public void registerLocation(Carrier carrier, Location location, Date time) {
-        refreshLastPositions(time,location);
-        gridInteractor.write(location, String.valueOf(time.getTime()), carrier, successListener, failureListener);
+        registerLocation(carrier, location, time, successListener, failureListener);
     }
+
     @Override
     public void registerLocation(Carrier carrier,
                                  Location location,
                                  Date time,
                                  OnSuccessListener successListener,
-                                 OnFailureListener failureListener){
-        refreshLastPositions(time,location);
-        gridInteractor.write(location, String.valueOf(time.getTime()), carrier, successListener, failureListener);
+                                 OnFailureListener failureListener) {
+        refreshLastPositions(time, location);
 
+        Map<String, Object> element = new HashMap<>();
+        element.put("geoPoint", new GeoPoint(location.getLatitude(), location.getLongitude()));
+        element.put("timeStamp", time.getTime());
+        element.put("infectionStatus", carrier.getInfectionStatus());
+        gridInteractor.writeDocumentWithID("LastPositions", AccountFragment.getAccount(getActivity()).getId(),
+                element, s -> {
+                    gridInteractor.write(location, String.valueOf(time.getTime()), carrier, successListener, failureListener);
+                },
+                failureListener);
     }
+
     /**
      * removes every locations older than UNINTENTIONAL_CONTAGION_TIME ms and adds a new position
      */
     private void refreshLastPositions(Date time, Location location) {
-        Date oldestDate = new Date(time.getTime()-InfectionAnalyst.UNINTENTIONAL_CONTAGION_TIME);
+        Date oldestDate = new Date(time.getTime() - InfectionAnalyst.UNINTENTIONAL_CONTAGION_TIME);
         lastPositions.headMap(oldestDate).clear();
-        if(location!=null){
-            lastPositions.put(time,location);
+        if (location != null) {
```

Comment:
Who cares ? xD

Reply:
auto refactor from the IDE ;) I wrote a ugly bloc of code and asked the IDE to refactor it for me. The rest of the file was collateral improvement :p

Reply:
Oh ! beeing juged by an IDE...